views/js/runner/plugins/tools/textToSpeech/textToSpeech.js

```diff
@@ -0,0 +1,374 @@
+                this.trigger('setSpeed', speed);
+            },
+
+            /**
+             * Volume
+             */
+            setVolume: function setVolume(e, value) {
+                volume = +value;
+
+                this._exec('setVolumeValue', volume);
+                this.trigger('setVolume', volume);
+            },
+
+            /**
+             * Click to pronounce
+             */
+            clickToSpeak: function clickToSpeak() {
+                options.enableClickToSpeak = !options.enableClickToSpeak;
+                this._exec('enableClickToSpeak', options.enableClickToSpeak);
+                this._exec('enableContinuousReading', !options.enableClickToSpeak); // continuous reading is off when click to speak is on
+                this.trigger('clickToSpeak');
+            }
+        }, {
+            // defaults
+        })
+        .setTemplate(tpl)
+        .on('init', function () {
+            var self = this;
+
+            speed = 40; // default speed
+            volume = 40; // default volume
+
+            // we have to mark some blocks as ignored to prevent TTS accessing it
+            $(options.ignoreEls.join(',')).each(function () {
+                $(this).attr('ignore', true);
+            });
+
+            require([options.toolbarUrl], function () {
+                var tss = self._get('texthelpSpeechStream');
+
+                require(["//" + tss.g_strServer + "/SpeechStream/v" + tss.g_strBuild + "/texthelpMain.js"], function () {
+                    self._init();
+                });
+            });
+        })
+        .init(config)
+        .on('render', function () {
+            var $this = this.getElement();
+
+            // prevents clicks from removing highlighted text
+            $this.on('mousedown', function (e) {
+                e.preventDefault();
+                return false;
+            });
+
+            // Action clicks
+            $this.find('.action').on('click', function (e) {
+                // prevents disabled actions from being triggered
+                if ($(this).hasClass('disabled')) {
+                    e.stopImmediatePropagation();
+                }
+
+                // hides settings menu when another action clicked
+                if (!$(this).closest('.settings').length) {
+                    $this.find('.settings').removeClass('active');
+                    $this.find('.settings > .settings-menu').hide();
+                }
+            });
+
+            // Show/hide settings menu
+            $this.find('.settings').on('click', function () {
+                $(this).toggleClass('active');
+                $this.find('.settings-menu').toggle();
+            });
+
+            // Settings menu
+            $this.find('.settings > .settings-menu')
+            .on('click', function (e) { // prevent child elements triggering a click on settings menu
+                if ($(this).closest('.settings-menu').length) {
```

Comment:
strange assertion... this condition looks useless as closest() will test the element itself and traversing up through its ancestors. And since the selector '.settings > .settings-menu' is already selecting .settings-menu, it is always true. Is there any specific reason for it being here ?

Reply:
Hah, yes. Good catch. I recall adding this as on an ancestor element, where `closest()` was necessary. Later changing the element of the click event while forgetting to remove `closest()`.